aarch64: Static_assert fails

Hello,
It compiles fine on x86_64 but fail static_assert on aarch64 (Nvidia Xavier). 
```
ubuntu@ubuntu-desktop:~/dev/trantor/build$ make
Scanning dependencies of target trantor
[  4%] Building CXX object CMakeFiles/trantor.dir/trantor/utils/AsyncFileLogger.cc.o
[  8%] Building CXX object CMakeFiles/trantor.dir/trantor/utils/ConcurrentTaskQueue.cc.o
[ 12%] Building CXX object CMakeFiles/trantor.dir/trantor/utils/Date.cc.o
[ 16%] Building CXX object CMakeFiles/trantor.dir/trantor/utils/LogStream.cc.o
/home/ubuntu/dev/trantor/trantor/utils/LogStream.cc: In member function ‘void trantor::LogStream::staticCheck()’:
/home/ubuntu/dev/trantor/trantor/utils/LogStream.cc:107:5: error: static assertion failed
     static_assert(kMaxNumericSize - 10 >
     ^~~~~~~~~~~~~
CMakeFiles/trantor.dir/build.make:101: recipe for target 'CMakeFiles/trantor.dir/trantor/utils/LogStream.cc.o' failed
make[2]: *** [CMakeFiles/trantor.dir/trantor/utils/LogStream.cc.o] Error 1
CMakeFiles/Makefile2:72: recipe for target 'CMakeFiles/trantor.dir/all' failed
make[1]: *** [CMakeFiles/trantor.dir/all] Error 2
Makefile:129: recipe for target 'all' failed
make: *** [all] Error 2
```
I originally thought it was cross-compilation issue but same error happens when compiling directly on Nvidia Xavier board.

--- a/trantor/tests/LoggerTest.cc
+++ b/trantor/tests/LoggerTest.cc
@@ -3,6 +3,12 @@
 #include <thread>
 int main()
 {
+    int i;
+    LOG_DEBUG << (float)3.14;
+    LOG_DEBUG << (const char)'8';
+    LOG_DEBUG << &i;
+    LOG_DEBUG << (long double)3.1415;
+    LOG_DEBUG << trantor::Fmt("%.3g", 3.1415926);
     LOG_DEBUG << "debug log!" << 1;
     LOG_TRACE << "trace log!" << 2;
     LOG_INFO << "info log!" << 3;

--- a/trantor/utils/LogStream.cc
+++ b/trantor/utils/LogStream.cc
@@ -100,23 +100,10 @@ void FixedBuffer<SIZE>::cookieEnd()
 {
 }
 
-void LogStream::staticCheck()
-{
-    static_assert(kMaxNumericSize - 10 > std::numeric_limits<double>::digits10,
-                  "");
-    static_assert(kMaxNumericSize - 10 >
-                      std::numeric_limits<long double>::digits10,
-                  "");
-    static_assert(kMaxNumericSize - 10 > std::numeric_limits<long>::digits10,
-                  "");
-    static_assert(kMaxNumericSize - 10 >
-                      std::numeric_limits<long long>::digits10,
-                  "");
-}
-
 template <typename T>
 void LogStream::formatInteger(T v)
 {
+    constexpr static int kMaxNumericSize = std::numeric_limits<T>::digits10 + 4;
     if (_exBuffer.empty())
     {
         if (_buffer.avail() >= kMaxNumericSize)
@@ -172,13 +159,13 @@ LogStream &LogStream::operator<<(unsigned long v)
     return *this;
 }
 
-LogStream &LogStream::operator<<(long long v)
+LogStream &LogStream::operator<<(const long long &v)
 {
     formatInteger(v);
     return *this;
 }
 
-LogStream &LogStream::operator<<(unsigned long long v)
+LogStream &LogStream::operator<<(const unsigned long long &v)
 {
     formatInteger(v);
     return *this;
@@ -187,6 +174,8 @@ LogStream &LogStream::operator<<(unsigned long long v)
 LogStream &LogStream::operator<<(const void *p)
 {
     uintptr_t v = reinterpret_cast<uintptr_t>(p);
+    constexpr static int kMaxNumericSize =
+        std::numeric_limits<uintptr_t>::digits / 4 + 4;
     if (_exBuffer.empty())
     {
         if (_buffer.avail() >= kMaxNumericSize)
@@ -214,8 +203,9 @@ LogStream &LogStream::operator<<(const void *p)
 }
 
 // TODO: replace this with Grisu3 by Florian Loitsch.
-LogStream &LogStream::operator<<(double v)
+LogStream &LogStream::operator<<(const double &v)
 {
+    constexpr static int kMaxNumericSize = 32;
     if (_exBuffer.empty())
     {
         if (_buffer.avail() >= kMaxNumericSize)
@@ -232,6 +222,29 @@ LogStream &LogStream::operator<<(double v)
     auto oldLen = _exBuffer.length();
     _exBuffer.resize(oldLen + kMaxNumericSize);
     int len = snprintf(&(_exBuffer[oldLen]), kMaxNumericSize, "%.12g", v);
+    _exBuffer.resize(oldLen + len);
+    return *this;
+}
+
+LogStream &LogStream::operator<<(const long double &v)
+{
+    constexpr static int kMaxNumericSize = 48;
+    if (_exBuffer.empty())
+    {
+        if (_buffer.avail() >= kMaxNumericSize)
+        {
+            int len = snprintf(_buffer.current(), kMaxNumericSize, "%.12Lg", v);
+            _buffer.add(len);
+            return *this;
+        }
+        else
+        {
+            _exBuffer.append(_buffer.data(), _buffer.length());
+        }
+    }
+    auto oldLen = _exBuffer.length();
+    _exBuffer.resize(oldLen + kMaxNumericSize);
+    int len = snprintf(&(_exBuffer[oldLen]), kMaxNumericSize, "%.12Lg", v);
     _exBuffer.resize(oldLen + len);
     return *this;
 }

--- a/trantor/utils/LogStream.h
+++ b/trantor/utils/LogStream.h
@@ -135,18 +135,18 @@ class LogStream : NonCopyable
     self &operator<<(unsigned int);
     self &operator<<(long);
     self &operator<<(unsigned long);
-    self &operator<<(long long);
-    self &operator<<(unsigned long long);
+    self &operator<<(const long long &);
+    self &operator<<(const unsigned long long &);
 
     self &operator<<(const void *);
 
-    self &operator<<(float v)
+    self &operator<<(float &v)
     {
         *this << static_cast<double>(v);
         return *this;
     }
-    self &operator<<(double);
-    // self& operator<<(long double);
+    self &operator<<(const double &);
+    self &operator<<(const long double &v);
 
     self &operator<<(char v)
     {
@@ -250,14 +250,11 @@ class LogStream : NonCopyable
     }
 
   private:
-    void staticCheck();
-
     template <typename T>
     void formatInteger(T);
 
     Buffer _buffer;
     std::string _exBuffer;
-    static const int kMaxNumericSize = 32;
 };
 
 class Fmt  // : boost::noncopyable
@@ -276,7 +273,7 @@ class Fmt  // : boost::noncopyable
     }
 
   private:
-    char _buf[32];
+    char _buf[48];
     int _length;
 };
 

